axiom/nr_segment.c

```diff
@@ -1170,23 +1170,59 @@ void nr_segment_record_exception(nr_segment_t* segment,
   nr_segment_set_error(segment, error_message, error_class);
 }
 
-void nr_segment_set_error(nr_segment_t* segment,
-                          const char* error_message,
-                          const char* error_class) {
-  if ((NULL == segment) || (NULL == error_message && NULL == error_class)) {
-    return;
-  }
-
+static void nr_segment_set_error_helper(nr_segment_t* segment,
+                                        const char* error_message,
+                                        const char* error_class,
+                                        const char* error_file,
+                                        int error_line,
+                                        char* error_context,
+                                        int error_no) {
   if (NULL == segment->error) {
     segment->error = nr_zalloc(sizeof(nr_segment_error_t));
   }
 
   nr_free(segment->error->error_message);
   nr_free(segment->error->error_class);
+  if (NULL != error_file && NULL != error_context) {
+    nr_free(segment->error->error_file);
+    nr_free(segment->error->error_context);
+  }
 
   segment->error->error_message
       = error_message ? nr_strdup(error_message) : NULL;
   segment->error->error_class = error_class ? nr_strdup(error_class) : NULL;
+  if (NULL != error_file && NULL != error_context) {
```

Comment:
I'm not clear on the behavior of this function if the segment had all fields defined previously and then this function is called and `error_file` and `error_context` are `NULL`.  It seems the previous values would be retained.  That could be confusing as it seems there is no way to clear those fields once set.

Reply:
After discussion, it has been decided that if `NULL` is passed in, the attribute should not show up. I will modify the code and create a commit for this.

Reply:
I have modified this function where all strings are freed now, so the previous values will not be retained. New values will only be added if they are not `NULL`: [ed191c0](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/ed191c0b4ee2cff587aa3fcee1219c9133e586f7), [bc0a82e](https://github.com/newrelic/newrelic-php-agent/pull/942/commits/bc0a82e066d75632bad2a84be5fa90b5dc42d5ed)